KAFKA-10126:Add a warning message for ConsumerPerformance

jiameixie opened the pull request:
ConsumerPerformance has not implemented options numThreadsOpt and numFetchersOpt as so far. This patch adds a warning message when used these options according to comments from https://issues.apache.org/jira/browse/KAFKA-10126 . Once these two options are implemented, this warning message should be removed.

Change-Id: Iae34b11210a361373eda42274ebfc153b636a989
Signed-off-by: Jiamei Xie <jiamei.xie@arm.com>

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

jiameixie committed:
KAFKA-10126:Warn unused ConsumerPerformance optons

Option numThreadsOpt and numFetchersOpt are unused in ConsumerPerformance.
According to comments from https://issues.apache.org/jira/browse/KAFKA-10126,
add a warning when invoking these options which should be recognized

Change-Id: Iae34b11210a361373eda42274ebfc153b636a989
Signed-off-by: Jiamei Xie <jiamei.xie@arm.com>

jiameixie commented:
@abbccdda @chia7712 @ijuma  Call for review. Thanks.

chia7712 commented on core/src/main/scala/kafka/tools/ConsumerPerformance.scala:
```diff
@@ -232,12 +232,12 @@ object ConsumerPerformance extends LazyLogging {
       .ofType(classOf[java.lang.Integer])
       .defaultsTo(2 * 1024 * 1024)
-    val numThreadsOpt = parser.accepts("threads", "Number of processing threads.")
+    val numThreadsOpt = parser.accepts("threads", "Number of processing threads. It has not been implemented.")
```
https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/tools/GetOffsetShell.scala#L55 had similar issue and we can follow its style. i.e ```DEPRECATED AND IGNORED:```

abbccdda replied:
+1

abbccdda commented on core/src/main/scala/kafka/tools/ConsumerPerformance.scala:
```diff
@@ -256,6 +256,10 @@ object ConsumerPerformance extends LazyLogging {
     options = parser.parse(args: _*)
+
+    if(options.has(numThreadsOpt) || options.has(numFetchersOpt))
+      println("WARNING: option threads and num-fetch-threads has not been implemented.")
```
`has not been implemented` -> `have been deprecated`

abbccdda reviewed (commented):
Overall LGTM plus @chia7712 comment.

jiameixie committed:
KAFKA-10126:Add a warning message for ConsumerPerformance

ConsumerPerformance has not implemented options numThreadsOpt and numFetchersOpt as so far.
This patch adds a warning message when used these options according to comments from
https://issues.apache.org/jira/browse/KAFKA-10126 . Once these two options are implemented,
this warning message should be removed.

Signed-off-by: Jiamei Xie <jiamei.xie@arm.com>

jiameixie commented:
@abbccdda @chia7712 Thanks for your advice. I have updated it.

abbccdda commented on core/src/main/scala/kafka/tools/ConsumerPerformance.scala:
```diff
@@ -256,6 +256,10 @@ object ConsumerPerformance extends LazyLogging {
     options = parser.parse(args: _*)
+
+    if(options.has(numThreadsOpt) || options.has(numFetchersOpt))
+      println("WARNING: option threads and num-fetch-threads have been deprecated and ignored")
```
nit: add square brackets to the options, like:
```
"WARNING: option [threads] and [num-fetch-threads] have been deprecated and will be ignored by the test"
```

abbccdda reviewed (commented):
Just a minor comment

jiameixie committed:
Change warning message

Signed-off-by: Jiamei Xie <jiamei.xie@arm.com>

abbccdda reviewed (approved):
LGTM, thanks for the patch!

jiameixie commented:
@ijuma Call for review

jiameixie commented:
@abbccdda @chia7712 Could you merge this PR please？Thanks

abbccdda commented:
test this please

abbccdda commented:
Only one flaky test failed:
```
kafka.api.PlaintextAdminIntegrationTest.testConsumerGroups
```

abbccdda commented:
retest this please

abbccdda commented:
retest this please

abbccdda commented:
Also failed: 
```
kafka.api.PlaintextAdminIntegrationTest.testAlterReplicaLogDirs

org.scalatest.exceptions.TestFailedException: only 0 messages are produced within timeout after replica movement. Producer future Some(Failure(java.util.concurrent.TimeoutException: Timeout after waiting for 10000 ms.))
	at org.scalatest.Assertions.newAssertionFailedException(Assertions.scala:530)
	at org.scalatest.Assertions.newAssertionFailedException$(Assertions.scala:529)
	at org.scalatest.Assertions$.newAssertionFailedException(Assertions.scala:1389)
	at org.scalatest.Assertions.fail(Assertions.scala:1091)
	at org.scalatest.Assertions.fail$(Assertions.scala:1087)
	at org.scalatest.Assertions$.fail(Assertions.scala:1389)
	at kafka.api.PlaintextAdminIntegrationTest.testAlterReplicaLogDirs(PlaintextAdminIntegrationTest.scala:289)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:288)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:282)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:830)
```

abbccdda commented:
Having tested 3 * 3 times, and only flaky test failed such as this one:
https://issues.apache.org/jira/browse/KAFKA-9509
